commands/eval.b.inc

```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * @file
+ * Command(s) for evaluating PHP code directly.
+ */
+
+/**
+ * Implements hook_b_command().
+ */
+function eval_b_command() {
+  return array(
+    'php-eval' => array(
+      'description' => bt('Evaluate arbitrary PHP code after bootstrapping Backdrop.'),
+      'callback' => 'php_eval_b_command',
+      'arguments' => array(
+        'code' => bt('The PHP code to evaluate.'),
+      ),
+      'aliases' => array('ev', 'eval'),
+      'bootstrap' => B_BOOTSTRAP_FULL,
```

Comment:
Is this really necessary? What if I wanted to check the PHP version installed on my system without having a Backdrop site setup yet? Including this means generic commands can't be run without a full Backdrop site, whereas removing it means they can, but doesn't prevent commands interacting with Backdrop when available (i.e. the best of both worlds). So I'd suggest removing this.

```suggestion
```

Reply:
As with the description, I just followed the setting that drush used. I'm fine with removing it; can't think of any reason (other than the description) why full bootstrapping would be needed.

Reply:
OK, after some further testing, it is clear that `B_BOOTSTRAP_FULL` is needed.

A common use case for this command is to wrap a bunch of PHP commands that you'd like to execute inside of a function in some module, then you invoke that function with the command `b ev 'my_module_something()'`. For this to work, however, you need Backdrop fully bootstrapped, so that specifically the code that you added to your module gets included.

For that matter, even if you're just calling a Backdrop API function, you need Backdrop bootstrapped at least far enough to bring that function into scope. In general, for most real use cases, you're going to want things fully bootstrapped. And I'm sure the use cases where you need fully bootstrapped outnumber the cases where you'd want to use this before things are fully bootstrapped.

And besides, that's how drush does it. I realize that b aspires to be better than drush, but if people are coming to b from drush, they'll likely have the reasonable expectation that commands with the same name behave the same way.

Reply:
Ok, I can see your point. For posterity, here's what I tested (all commands are run from *outside* the Backdrop installation):

**Bootstrap required:**
- `b eval "echo 'Testing 123'"` -> ERROR: "The required bootstrap level for 'php-eval' is not ready."
- `b eval "echo 'Testing 123'" --root=backdrop` -> Outputs 'Testing 123' as expected
- `b eval "backdrop_cron_run()"` -> ERROR: "The required bootstrap level for 'php-eval' is not ready."
- `b eval "backdrop_cron_run()" --root=backdrop` -> Cron runs successfully (no output)

**Bootstrap *not* required:**
- `b eval "echo 'Testing 123'"` -> Outputs 'Testing 123' as expected
- `b eval "echo 'Testing 123'" --root=backdrop` -> Outputs 'Testing 123' as expected
- `b eval "backdrop_cron_run()"` -> ERROR: "PHP Fatal error:  Uncaught Error: Call to undefined function backdrop_cron_run() in /app/commands/eval.b.inc(34) : eval()'d code:1"
- `b eval "backdrop_cron_run()" --root=backdrop` -> ERROR: "PHP Fatal error:  Uncaught Error: Call to undefined function backdrop_cron_run() in /app/commands/eval.b.inc(34) : eval()'d code:1"

So while requiring bootstrap means you can't run commands outside of Backdrop, you can at least still run them all inside Backdrop (or by specifying the Backdrop installation using `--root`).

However if we remove the bootstrap requirement like I was suggesting, then it's impossible to call Backdrop functions either outside or inside a Backdrop install. Since that's a bigger problem, then I see the need for this now.